### PR TITLE
resolve  #612, #611 (2-step join w/ facebook, google)

### LINKED ISSUE #611: join group in 2 steps with google [3]
# Acceptance Criteria

GIVEN I am a not-logged-in user viewing a group's join page (`groups/1/join`)
* THEN I should see a button that says "Join with Facebook"
* WHEN I click on it I should go to a new member page (`groups/1/members/new`)
* THEN the page should display my first name, last name, and email (as reported by google)
* AND the page should have fields for my zip (required) and phone (optional)
* AND the system should *NOT* yet create a new user

GIVEN I have filled out the form without error
* WHEN I click "SUBMIT" 
* THEN the system should create a new user
* AND I should be logged in and redirected to `/home`

cursor_and_affinity

### LINKED ISSUE #612: join group in 2 steps with facebook [3]
# Acceptance Criteria

GIVEN I am a not-logged-in user viewing a group's join page (`groups/1/join`)
* THEN I should see a button that says "Join with Facebook"
* WHEN I click on it I should go to a new member page (`groups/1/
* THEN the page should display my first name, last name, and email (gleaned from facebook)
* AND the page should have fields for my zip (required) and phone (optional)
* AND the system should *NOT* yet create a new user

GIVEN I have filled out the form without error
* WHEN I click "SUBMIT" 
* THEN the system should create a new user
* AND I should be logged in and redirected to `/home`

### PR DESCRIPTION
**Pull request by aguestuser**
resolves #612 
resolves #611 

_________________________________

# notes

* provide 2 step flow for joining a group with either FB or Google OAuth
* do not create a person or store an identity upon OAuth callback (because all required fields have not yet been filled out)
* instead, use RbNaCl to symmetrically encrypt the OAuth token and pass it as a param to `members#new` and decrypt it before persisting
* support long-lived access tokens for FB join path, leave them unimplemented for Google
* add browser tests for login
* leaving this for both of you to review, as it's fairly meaty

# Screenshot

oath-join-screencap

### CODE REVIEW

**Comment by aguestuser** [comment]
UPDATE:

i unborked my system and fixed the really annoying deprecation warnings!

here is a screenshot of what it looks like:

oath-join-screencap

here is a screengrab of JUST how annoying the warnings were:

super-annoying-deprecation-warning

here is a screengrab of what the tests now look like:

nice-tests

File: app/models/identity.rb [hunk 1]
```diff
@@ -1,23 +1,57 @@
...
+#  updated_at   :datetime         not null
+#  access_token :string
+#
+
 class Identity < ActiveRecord::Base
   before_save :request_long_lived_token
 
   belongs_to :person
-  validates_presence_of :uid, :provider, :person_id
+  validates_presence_of :uid, :provider#, :person_id
```
Comment (vaughanj10) [inline app/models/identity.rb]: Did you mean to leave person_id commented out?

**Comment by aguestuser** [comment]
UPDATE: fixed the problem with the nasty deprecation warnings...

File: app/models/identity.rb (hunk 1 above)
Comment (aguestuser) [reply app/models/identity.rb]: yes, but i should be more explicit about why. (it's in a commit message somewhere.)

the reason is that you can't pass `Identity` as a nested attribute to `person` if it validates the presence of `person_id`. for it to be a nested attribute, you have to be able to construct it with `person_id` as nil, and then give it `person_id` once the person it is nested inside of has been saved.

**Comment by Matt-Holland** [comment]
merging to master for testing

**Pull request merged**